Path: Fix roughly equal depth entries

Russ4262 opened the pull request:
Add code block to remove first of any two consecutive, roughly equal depth values (twin values).

Add two unit tests to support new code block and use cases.

This PR addresses an error in the `PathUtils.depth_params()` class that occurs when the total depth is roughly equal to a single stepdown.

Forum identification at [G-code generated twice. Why?](https://forum.freecadweb.org/viewtopic.php?f=15&t=67168&sid=818fe4ba0b7055ab736084b469599dd0).

Steps to reproduce:
* Create job with defaults.
* Create operation.
* Set Step Down value equal to total depth of cut (OpStartDepth - OpFinalDepth).
* Save operation.
* Inspect G-code for same operation, and note line count.  (This will produce one instance of the g-code.)
* Set Start Depth to say 0.00000001 mm more than existing entry (OpStartDepth + 0.00000001 mm).
* Save operation.
* Inspect G-code for same operation, and note increased line count.  (This will produce two stepdowns that are only about 0.00000001 mm  different - basically a copy of the single instance.)

Noting the difference in g-code line counts is easier with Adaptive because it throws in some comment lines as communication for the user.  However, this error exists with the other ops as well due to the use of the `PathUtils.depth_params()` class.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [x]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

sliptonic commented:
Nice catch, Russ!   The depthparams code has been so stable for so long, I'm surprised we haven't seen this before.
Since this code is very stable, it's worth spending some additional effort at readability and efficiency.  I've added some review items to consider.

Russ4262 commented:
> ... I've added some review items to consider.

I am not finding them.

sliptonic commented:
> 

I guess I gotta hit the 'I'm done with the code review button'.  Sorry

Russ4262 commented:
@sliptonic 
Okay.  Reworked the code a bit, and extended the roughly-equal filter to user-provided depths via the `user_depths` argument. Added a few more unit tests to verify the filter works as intended.

I think it is ready for round two of the review process.

Thanks.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6592/pipeline.svg"/></a> for feature branch [PR_6592](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592). Pipeline [494383394](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/494383394) was triggered at [6d5fc83](https://github.com/FreeCAD/FreeCAD/pull/6592/commits/6d5fc83951ca1f8c0fa049b21b37fc94be3112a8). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

Russ4262 commented:
@sliptonic I believe this is ready for final review.

@mlampert Will you please provide feedback about the change to unit test 003 (previously 30).  This PR adds a descending sort to `user_depths` provided in the `depth_params` class, via a new filter method added; therefore, I changed the expected result in the unit test to reflect the applied descending sort added in this PR.  My thought is that `depth_params` should be consistent with descending results, until we decide to add a `reverse` attribute or argument for later use with additive construction process.

Thanks for your time, feedback and coding instruction.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6592/pipeline.svg"/></a> for feature branch [PR_6592](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592). Pipeline [496115013](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/496115013) was triggered at [7159723](https://github.com/FreeCAD/FreeCAD/pull/6592/commits/7159723283cbe701ac2738ef0ad1667a6371bfe2). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

sliptonic commented on src/Mod/Path/PathScripts/PathUtils.py:
```diff
@@ -718,7 +722,30 @@ def __get_depths(self, equalstep=False):
+        of repetitive roughly-equal values."""
+
+        # Do not need a copy of original `depths` because we are only reading and copying
+        depths.sort()  # Sort again in case user_depths were not sorted
```
A good practice is to NEVER modify directly the arguments passed into the function. 
sort() changes the depths list.  It actually modifies the current list.  Doing so can cause unexpected consequences, especially if the code is later modified to work in a multi-threaded way.  This is one of the bedrock principles of functional programming and while we aren't trying to write functional code, it's an easy thing to implement that makes the code more reliable.

This case is low risk, but you could imagine that another thread is simultaneously processing the list for another purpose.  Sorting the list here could break the other thread and it would be very difficult to debug because the bug would not be consistent.

mlampert commented on src/Mod/Path/PathScripts/PathUtils.py:
```diff
@@ -718,7 +722,30 @@ def __get_depths(self, equalstep=False):
+            if not PathGeom.isRoughly(depth, keep[-1]):
+                keep.append(depth)  # if different, add to keep
+        keep.reverse()  # reverse results back high to low
+        return keep
```
sliptonic's comment got me thinking on how to concisely do this. Note that I'm mostly writing this down so I'll remember it, it's not meant as an invitation for refactoring
```
def __filter_roughly_equal_depths(depths):
  last = None
  def unique(depths):
    for depth in sorted(depths):
      if last is None or not PathGeom.isRoughly(depth, last):
        last = depth
        yield i
  return list(unique(depths))
```

sliptonic replied:
Clever.  I've not used sorted().  It does the copy and sorting in one step without modifying the original. Nice.

@Russ4262 Feel free to kill the overly detailed comments.  That was just for our discussion.

mlampert reviewed (approved):
This is great!

Russ4262 commented:
Thanks Sliptonic and MLampert for the reviews.  I just pushed changes per your comments.  I reverted my changes to the `user_depths` attribute, leaving it untouched. This should be ready for final review.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6592/pipeline.svg"/></a> for feature branch [PR_6592](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592). Pipeline [496447118](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/496447118) was triggered at [30ca1a4](https://github.com/FreeCAD/FreeCAD/pull/6592/commits/30ca1a4f3e30a72ebcd780bae0dda99b00e1192b). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

Russ4262 commented:
Corrections made per comments above.  Also rebased on master.  Thanks.

freecadci commented:
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592"><img alt="pipeline status" src="https://gitlab.com/freecad/FreeCAD-CI/badges/PR_6592/pipeline.svg"/></a> for feature branch [PR_6592](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6592). Pipeline [496776185](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/496776185) was triggered at [95cffaf](https://github.com/FreeCAD/FreeCAD/pull/6592/commits/95cffaf502b0a9d2016ae6d8234bc2d6789305c4). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).
